secboot/secboot.go

```diff
@@ -100,6 +100,13 @@ type SealKeysParams struct {
 	TPMLockoutAuthFile string
 	// Whether we should provision the TPM
 	TPMProvision bool
+	// Only meaningful when used in conjunction with TPMProvision. Assumes a
+	// partial reprovisioniong of the TPM which was previously already
+	// provisioned by secboot. Existing lockout authorization data from
+	// TPMLockoutAuthFile will be used to authorize provisioning and will
+	// get overwritten in the process. The resource handles identified by
+	// PCRPolicyCounterHandle will be released before allocating them again
```

Comment:
I'm confused about how this is meant to be used. The auth key is shared between all sealed keys and needs to be rotated in a factory reset, so SealKeys needs to be called for both the run-mode and fallback sealed keys (for both ubuntu-data and ubuntu-save). In both cases, you will need to delete the existing PCR policy counter but you'll only want to actually reprovision once, which implies that there needs to be a way to seal new keys and delete an existing PCR policy counter with `!TPMProvision`.

Also, this new field is being used to control the behaviour for both a device-level action (changing the TPMs lockout authorization value) and a key-level action (deleting a PCR policy counter associated with an existing key).

Reply:
This follows the approach of the original UC20 sealing code, which is when sealing the run object key, TPMProvision is set to true and the authorization key gets written to a file. Though, when called right after to seal the fallback object key TPMProvision is false, and the authorization key is read back from the file. By extension, TPMPartialReprovision will also be true in the first call, but false in the second.

FWIW about mixing the responsibility, it's something I don't like either and find the snapd/secboot wrapper a bit unforunate in this regard. Refactoring would be nice, but I need to discuss that with @pedronis as to whether do it now, or once we're done with factory (or at all).

Reply:
I think the API was designed when it wasn't clear we would get all this kind of different keys. I  think refactoring would be maybe a good idea but can be done later.  Having the two bool flags is still strange though, maybe they should become an a single enum fields instead?